samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/SubmitToGradingActionListener.java

```diff
@@ -814,18 +814,43 @@ private void prepareItemGradingPerItem(ActionEvent ae, DeliveryBean delivery,
 			answerModified = false;
 			for (int m = 0; m < grading.size(); m++) {
 				ItemGradingData itemgrading = grading.get(m);
-				if (itemgrading != null) {
-					log.debug("\n:ItemId>>itemTextId>>answerId "+ itemgrading.getPublishedItemId()+itemgrading.getPublishedItemTextId()+itemgrading.getPublishedAnswerId()+"\n");	    
-
-					itemgrading.setAgentId(AgentFacade.getAgentString());
-					itemgrading.setSubmittedDate(new Date());
-					if (itemgrading.getRationale() != null && itemgrading.getRationale().length() > 0) {
-						itemgrading.setRationale(TextFormat.convertPlaintextToFormattedTextNoHighUnicode(log, itemgrading.getRationale()));
+				if ((itemgrading !=null && itemgrading.getItemGradingId() == null) || itemgrading.getItemGradingId().intValue() <= 0) {
+					if (itemgrading.getPublishedAnswerId() != null || (itemgrading.getRationale() != null && !itemgrading.getRationale().trim().equals(""))) { 
+						answerModified = true;
+						break;
 					}
-					adds.add(itemgrading);
 				}
 			}
+			// Click the Reset Selection link
+			if(item.getUnanswered()) {
+				answerModified = true;
+			}
 
+			if (answerModified) {
+				for (int m = 0; m < grading.size(); m++) {
+					ItemGradingData itemgrading = grading.get(m);
+					if (itemgrading !=null && itemgrading.getItemGradingId() != null && itemgrading.getItemGradingId().intValue() > 0) {
+						// remove all old answer
+						removes.add(itemgrading);
+					} else {
+						// add new answer
+						if (itemgrading.getPublishedAnswerId() != null
```

Comment:
Findbugs pointed this out to be a possible NPE. You're checking it in the if case above but not in this if case as part of the else. Can it be null?